sbol2/document.py

```diff
@@ -447,39 +447,76 @@ def writeString(self):
+        new_graph.parse(filename, format='application/rdf+xml')
+        self._append_graph(new_graph, overwrite)
+
+        # self.logger.debug("Appending data from file: " + filename)
+        # # Write our SBOL objects to an RDFlib graph
+        # self.update_graph()
+        # # Use rdflib to automatically merge the graphs together
+        # self.graph.parse(filename, format="application/rdf+xml")
+        # # Clear out the SBOL objects, but not the newly merged graph
+        # self.clear(clear_graph=False)
+        # # Base our internal representation on the new graph.
+        # self.parse_all()
+
+    def appendString(self, sbol_str: str, overwrite=True):
         """
         Read an RDF/XML document from a string and attach the SBOL
         objects to this Document.
 
         New objects will be added to the existing contents of the Document.
         :param sbol_str: A string of RDF/XML
+        :param overwrite: Boolean indicating whether to overwrite existing objects
         :return: None
         """
-        # Save any changes we've made to the graph.
-        self.update_graph()
-        # Use rdflib to automatically merge the graphs together
-        self.graph.parse(data=sbol_str, format="application/rdf+xml")
-        # Base our internal representation on the new graph.
+        # ------------------------------------------------------------
+        # Load the new data into a graph
+        new_graph = rdflib.Graph()
+        new_graph.parse(data=sbol_str, format='application/rdf+xml')
+        self._append_graph(new_graph, overwrite)
+
+    def _append_graph(self, new_graph: rdflib.Graph, overwrite: bool = True):
+        # Gather all the objects that will be overwritten, stopping
+        # if the user says not to overwrite. If we clear as we go we lose
+        # the ability to find objects within objects. So gather the list
+        # here, and clear them as a second pass.
+        objects_to_clear = []
+        for s, _, _ in new_graph.triples((None, rdflib.RDF.type, None)):
+            existing_object = self.find(s)
+            if existing_object is not None:
+                if overwrite is False:
+                    return False
+                objects_to_clear.append(existing_object)
+        # Clear out the internal stores of the objects_to_clear so that
+        # they will be overwritten. Keep the identity property because it
+        # does not get restored by the graph parsing. Keep all the keys in
+        # the internal stores so owned objects end up in the right place.
+        for existing_object in objects_to_clear:
+            if existing_object.identity not in self.SBOLObjects:
+                self.SBOLObjects[existing_object.identity] = existing_object
+            # Now clear the object. It will get reconstituted below by parse_all
+            for k in existing_object.properties:
+                # Clear out everything but the identity
+                if k == SBOL_IDENTITY:
```

Comment:
Is this actually necessary?  Seems like we've already confirmed that the identities are equal, so what's the harm in just clearing everything?

Reply:
There is no way to have identity set if we take out this line. `identity` isn't actually serialized to an SBOL file, and on reading from RDF the identity is set when the object is created. Since the object has already been created we have to leave the identity in place.

Reply:
OK no biggie